File: 20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> match;
        for (int i = 0; i < s.size(); ++i)
        {
            if (!match.empty() && ((s[i] == ')' && match.top() == '(') || (s[i] == '}' && match.top() == '{') || (s[i] == ']' && match.top() == '[')))
            {
                match.pop();
            }
            else
            {
                match.push(s[i]);
            }
        }
        return match.size() ? false : true;
    }
};
// @lc code=end

File: JavaScript/20.有效的括号.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = new Set();
  stack.add(")");
  stack.add("}");
  stack.add("]");

  const match = new Map();
  match.set(")", "(");
  match.set("}", "{");
  match.set("]", "[");

  const stackArray = [];

  for (let i = 0; i < s.length; ++i) {
    const char = s[i];

    if (stack.has(char)) {
      if (
        stackArray.length &&
        stackArray[stackArray.length - 1] === match.get(char)
      ) {
        stackArray.pop();
      } else {
        return false;
      }
    } else {
      stackArray.push(char);
    }
  }

  return stackArray.length === 0;
};
// @lc code=end
